### ISSUE TITLE
A better logging...

### ISSUE THREAD

**Issue by jdaniele71** [firedm/FireDM#314]
Hi @firedm ,

I would ask you to add a real log to FireDM.

As far I can see by a quick code review, at the moment it prints out messages on standard out and, with a call_back function, it appends them to the log tab, without any timestamp.

As you can see on issue #303 the log tab is a bit problematic (probably it is not able to manage thousands of lines, probably we should remove older lines before appending new ones when it reaches the 100000 chars limit  you set), but even if it would perfect, it cannot manage application crashing or freezing. As soon as we close the app, everything disappears.

I think we definitely need a log file! 😅

Python comes with his nice logging package, as you know.
Why don't we use it?
I was reading this nice article about logging [https://blog.guilatrova.dev/how-to-log-in-python-like-a-pro/](url) , because i love to learn any better way to log.
When there is any issue in production environment, on customer premise, what really saved my ass many times was a fairly verbose and smart, file based logging! Period! 😋

It is since 1997 that i take so much care of logging.
If i find a production application without logging that I'm called to work on,  the very FIRST thing I do on the app is to add logging, to any class, to nearly any function call (with some exception). 
It doesn't matter if I have to modify 50 classes and 500 methods, I did it many times... 😁

Honestly, I have not so much time but, I would like to give my contribution to your great project if you like too.
I mean, I would help you adding logging, so I will be forced to learn the code and know the application better.
In the future I could maybe help you to fix something or add some requested enhancements, as far as i have time, as far as you like.
I think the time I spend writing "issues" on github, could be better employed writing/fixing code... 😂

What do you think about it?
PS: I won't be offended if you don't want it, It's your project...no problem at all! 👍

Thanks
Daniele

**Comment by firedm** [comment]
@jdaniele71 
you're welcome to participate in code anytime, just make sure you're changes is small, like a bug fix or add feature which is relatively small that affect one file or 2 as a max 
in this way merging changes won't produce a conflict with currently ongoing modifications by others
  
regarding log sys, definitely it needs enhancement, but changing log system will require modifying all files in the project, which is a time consuming task.

**Comment by jdaniele71** [comment]
> @jdaniele71 you're welcome to participate in code anytime, just make sure you're changes is small, like a bug fix or add feature which is relatively small that affect one file or 2 as a max in this way merging changes won't produce a conflict with currently ongoing modifications by others
> 
> regarding log sys, definitely it needs enhancement, but changing log system will require modifying all files in the project, which is a time consuming task.

@firedm 
Yes, it is a time consuming task, but i did it hundreds of times (like a robot 🦾) on (C#) projects having hundreds of methods.
By the way, the two logging systems can coexist for a while, during the transition.
There are more or less 15-16 python source code files where logging can be added/modified.
In the first pass, I could add the logging package to project and use it in just one file, without removing the old log.
They will log in parallel, one as usual, the other on disk.
Than we could choose to merge or to continue adding the log to other files.

I usually follow this schema:

```
method definition(parameters.....):
   logging method entering
   defining the result variable
   try
      raise exceptions for bad parameters(not respecting the contract)
      logging most interesting parameters
      doing the actual method job

      logging the condition
      if condition:
        do something
      setting the result
   except
      logging the error
   finally
      logging the result (if any)
      logging function exiting
   return the result
```
Notice as there is only one return and there is an explicit declaration of the result variable(s).
I don't like many returns where results have been built like "return a + b  + c" because I won't be able to log the result.

So when you read the log, you can know very closely what is happening or it happened inside the app.

Another faster and simpler approach is to add just below the "print" statement "print(start + text + end, end='', **kwargs)" in the utils.py a line with something like "logger.Debug(....)".
So in this very first introduction, adding few lines of code to import, configure the logging system and log from the "centralized" method, we'll have a very raw log on the disk.
We have just to agree on where to log, log rotation, logging fields....

About conflicts, having already configured the git upstream of my fork pointing to your main project, i can daily pull changes from it and resolve conflicts constantly and locally, pushing them to my fork.
So when the moment to open a pull request will arrive, all the files will be already aligned and no conflicts will appear.
Do you agree?

I have a question: how do you coordinate the development between developers?
For example, you said i can take an issue, and fix it.
How can i know if you or someone else are already working in it?
If i don't ask you, i risk to waste the time fixing something already fixed...😅
I think it would be a good and useful practice to ask you for "the go" before starting to work on specific issue/enhancement.
Am i wrong?

Thanks
Daniele

**Comment by firedm** [comment]
> Another faster and simpler approach is to add just below the "print" statement "print(start + text + end, end='', **kwargs)" in the utils.py a line with something like "logger.Debug(....)".
So in this very first introduction, adding few lines of code to import, configure the logging system and log from the "centralized" method, we'll have a very raw log on the disk.

sound good, you can start with this, then gradually modify the rest

> We have just to agree on where to log, log rotation, logging fields....

we store all firedm config files in setting folder, normally in current firedm folder if it has write access, otherwise it uses ~/.config/FireDM on linux or %APPDATA%/.FireDM on Windows, this will be same place for logging

> How can i know if you or someone else are already working in it?
If i don't ask you, i risk to waste the time fixing something already fixed...sweat_smile
> I think it would be a good and useful practice to ask you for "the go" before starting to work on specific issue/enhancement.
Am i wrong?

that's right, just let me know before working on something please, by either creating a new issue or commenting on existing issue, this way you won't waste your effort

**Comment by jdaniele71** [comment]
> sound good, you can start with this, then gradually modify the rest

Ok, fine, thanks. I will.

> we store all firedm config files in setting folder, normally in current firedm folder if it has write access, otherwise it uses ~/.config/FireDM on linux or %APPDATA%/.FireDM on Windows, this will be same place for logging

Ok, i see, thanks
 
> that's right, just let me know before working on something please, by either creating a new issue or commenting on existing issue, this way you won't waste your effort

Sure I will. I think it should be added as an advice in the FireDM developer guide.

Thanks
Daniele
